ouroboros-consensus/tools/db-analyse/Main.hs

```diff
@@ -0,0 +1,263 @@
+{-# LANGUAGE RankNTypes       #-}
+{-# LANGUAGE RecordWildCards  #-}
+
+module Main (main) where
+
+import           Control.Monad.Except
+import           Data.Coerce
+import           Data.Foldable (asum)
+import           Data.IORef
+import           Data.List (foldl', intercalate)
+import           Data.Reflection
+import           Data.Word
+import           Options.Applicative
+
+import           Cardano.Binary (unAnnotated)
+import qualified Cardano.Chain.Block as Chain
+import qualified Cardano.Chain.Genesis as Genesis
+import           Cardano.Chain.Slotting (EpochSlots (..))
+import qualified Cardano.Chain.UTxO as Chain
+import qualified Cardano.Crypto as Crypto
+
+import           Ouroboros.Network.Block (HasHeader (..), SlotNo (..),
+                     genesisPoint)
+
+import           Ouroboros.Consensus.Ledger.Byron (ByronBlockOrEBB, ByronGiven,
+                     ByronHash)
+import qualified Ouroboros.Consensus.Ledger.Byron as Byron
+import           Ouroboros.Consensus.Ledger.Byron.Config (ByronConfig)
+import           Ouroboros.Consensus.Node.Run.Abstract
+import           Ouroboros.Consensus.Node.Run.Byron ()
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Ouroboros.Storage.ChainDB.API (StreamFrom (..))
+import           Ouroboros.Storage.ChainDB.Impl.ImmDB
+import           Ouroboros.Storage.Common
+import           Ouroboros.Storage.EpochInfo
+import qualified Ouroboros.Storage.ImmutableDB.API as ImmDB
+--import           Ouroboros.Storage.ImmutableDB.Layout
+
+main :: IO ()
+main = do
+    CmdLine{..}   <- getCmdLine
+    genesisConfig <- openGenesis clConfig clIsMainNet
+    withByronGiven genesisConfig $ do
+      let epochInfo = fixedSizeEpochInfo (coerce (given :: EpochSlots))
+      immDB <- openImmDB clImmDB epochInfo
+      withRegistry $ runAnalysis clAnalysis immDB epochInfo
+    putStrLn "Done"
+
+{-------------------------------------------------------------------------------
+  Run the requested analysis
+-------------------------------------------------------------------------------}
+
+data AnalysisName =
+    ShowSlotBlockNo
+  | CountTxOutputs
+
+type Analysis = ByronGiven
+             => ImmDB IO Blk
+             -> EpochInfo IO
+             -> ResourceRegistry IO
+             -> IO ()
+
+runAnalysis :: AnalysisName -> Analysis
+runAnalysis ShowSlotBlockNo = showSlotBlockNo
+runAnalysis CountTxOutputs  = countTxOutputs
+
+{-------------------------------------------------------------------------------
+  Analysis: show block and slot number for all blocks
+-------------------------------------------------------------------------------}
+
+showSlotBlockNo :: Analysis
+showSlotBlockNo immDB _epochInfo rr =
+    processAll immDB rr go
+  where
+    go :: Either EpochNo SlotNo -> Blk -> IO ()
+    go isEBB blk = putStrLn $ intercalate "\t" [
+                       show isEBB
+                     , show (blockNo   blk)
+                     , show (blockSlot blk)
```

Comment:
You know I prefer `condense` instead of `show`+ a header line :grin:

Reply:
It makes the output less ambiguous. We'd have to do some additional massaging to clearly mark EBBs then. Let's put in the time when that's useful.